example/router/router.hpp

```diff
@@ -0,0 +1,260 @@
+
+        const_reference
+        operator[]( string_view id ) const;
+
+        const_iterator
+        find( string_view id ) const;
+
+        const_iterator
+        begin() const;
+
+        const_iterator
+        end() const;
+
+        bool
+        empty() const noexcept;
+
+        size_type
+        size() const noexcept;
+    };
+
+protected:
+    BOOST_URL_DECL
+    router_base();
+
+    BOOST_URL_DECL
+    void
+    route_impl(
+        string_view s,
+        std::size_t idx,
+        std::size_t maxn);
+
+    BOOST_URL_DECL
+    node const*
+    match_impl(
+        string_view s,
+        string_view*& matches,
+        string_view*& names) const noexcept;
+};
+
+
+/** A URL router.
+
+    This container matches static and dynamic
+    URL requests to an object which represents
+    how the it should be handled. These
+    values are usually callback functions.
+
+    @tparam T type of resource associated with
+    each path template
+
+    @tparam N maximum number of replacement fields
+    in a path template
+
+    @par Exception Safety
+
+    @li Functions marked `noexcept` provide the
+    no-throw guarantee, otherwise:
+
+    @li Functions which throw offer the strong
+    exception safety guarantee.
+
+    @see
+        @ref parse_absolute_uri,
+        @ref parse_relative_ref,
+        @ref parse_uri,
+        @ref parse_uri_reference,
+        @ref resolve.
+*/
+template <class T, std::size_t N = 20>
+class router : public router_base
+{
+public:
+    /// A range type with the match results
+    class match_results
+        : public match_results_base
+    {
+        string_view matches_storage_[N];
+        string_view ids_storage_[N];
+        T const* data_{nullptr};
+
+        friend router;
+
+        match_results() = default;
+
+        match_results(
+            T const* data,
+            node const* leaf,
+            string_view matches[N],
+            string_view ids[N],
+            std::size_t n)
+            : match_results_base(
+                leaf, matches_storage_, ids_storage_, n)
+            , data_(data)
+        {
+            for (std::size_t i = 0; i < n; ++i)
+            {
+                matches_storage_[i] = matches[i];
+                ids_storage_[i] = ids[i];
+            }
+        }
+
+    public:
+        T const&
+        operator*() const;
+
+        T const&
+        value() const;
+    };
+
+    /// Constructor
+    router() = default;
+
+    /** Route the specified URL path to a resource
+
+        @param path A url path with dynamic segments
+        @param resource A resource the path corresponds to
+
+        @see
+            https://fmt.dev/latest/syntax.html
+     */
+    template <class U>
+    void
+    route(string_view path, U&& resource);
+
+    /** Match URL path to corresponding resource
+
+        @param request Request path
+        @return The match results
+     */
+    match_results
+    match(string_view request) const noexcept;
```

Comment:
`request`? No its not a request it is just a string or a path, or something like that. Maybe `pct_string_view` or even `urls::encoded_segments_view`?

Reply:
> request? No its not a request it is just a string or a path, or something like that. 

Yes. It's more like just a `path` or `target`. `request` would be confusing because it sounds like `http::request` or something more complex.

> Maybe pct_string_view or even urls::encoded_segments_view?

It used to be `pct_string_view` but it didn't make much sense because we have to reparse it to get the segments. At this point, if the request is invalid, we simply don't have a match.

Reply:
what about `encoded_segments_view`?

Reply:
Yes. That's kind of what it does with the path. I'll check if there are implicit conversions, etc.